cmd/recorder.go

```diff
@@ -0,0 +1,308 @@
+	if sampleRate <= 0 || bufferSize <= 0 {
+		return
+	}
+	if len(FrameBuffer) >= sampleRate*FileDuration/bufferSize {
+		go flush(FrameBuffer, sampleRate)
+		FrameBuffer = []frame.Frame{}
+	}
+	subframes := make([]*frame.Subframe, NumRecorderChannels)
+	for i := range subframes {
+		subframe := &frame.Subframe{
+			Samples: make([]int32, bufferSize),
+		}
+		subframes[i] = subframe
+	}
+	// Dump raw sample data into respective subframes
+	for i, samples := range audioSamples {
+		subHdr := frame.SubHeader{
+			Pred:   frame.PredVerbatim,
+			Order:  0,
+			Wasted: 0,
+		}
+		subframes[i].SubHeader = subHdr
+		subframes[i].NSamples = bufferSize
+		subframes[i].Samples = subframes[i].Samples[:bufferSize]
+		for j, sample := range samples {
+			subframes[i].Samples[j] = int32(uint16(sample * math.MaxInt16))
+		}
+	}
+	// Package subframes into frame using stereo settings
+	header := frame.Header{
+		HasFixedBlockSize: false,
+		BlockSize:         uint16(bufferSize),
+		SampleRate:        uint32(sampleRate),
+		Channels:          frame.ChannelsLR,
+		BitsPerSample:     BitDepth,
+	}
+	frame := &frame.Frame{Header: header, Subframes: subframes}
+	FrameBuffer = append(FrameBuffer, *frame)
+}
+
+// createNewFLAC creates a new FLAC file based on server ID and current time
+func createNewFLAC() (*os.File, error) {
+	name := fmt.Sprintf("%s/raw-%s-%d.flac", MediaDir, os.Getenv("JACKTRIP_SERVER_ID"), time.Now().Unix())
+	fh, err := os.Create(name)
+	if err != nil {
+		return nil, err
+	}
+	return fh, nil
+}
+
+// openFLAC creates a FLAC encoder to write audio samples to a file
+func openFLAC(sampleRate int) (*flac.Encoder, error) {
+	fh, err := createNewFLAC()
+	if err != nil {
+		return nil, err
+	}
+	// Keep track of files created and rotate files
+	AudioFilenames = append(AudioFilenames, fh.Name())
+	if len(AudioFilenames) > FileCountLimit {
+		cleanStaleFile(AudioFilenames[0])
+		AudioFilenames = AudioFilenames[1:]
+	}
+	info := &meta.StreamInfo{
+		BlockSizeMin:  16,
+		BlockSizeMax:  65535,
+		SampleRate:    uint32(sampleRate),
+		NChannels:     NumRecorderChannels,
+		BitsPerSample: BitDepth,
+	}
+	return flac.NewEncoder(fh, info)
+}
+
+// cleanStaleFile deletes all files that pattern-match the input filename (minus extension)
+func cleanStaleFile(filename string) {
+	prefix := pathutil.TrimExt(filename)
+	files, _ := filepath.Glob(prefix + "*")
+	for _, f := range files {
+		os.Remove(f)
+	}
+}
+
+func updateHLSPlaylist() {
+	if HLSPlaylistHash != "" {
+		inputFile := AudioFilenames[len(AudioFilenames)-1]
+		basename := filepath.Base(inputFile)
+		basenameWithoutExt := strings.TrimSuffix(basename, filepath.Ext(basename))
+		// Execute ffmpeg - all options described here: https://ffmpeg.org/ffmpeg-formats.html
+		cmd := exec.Command(
+			// Call ffmpeg on the most-recently created FLAC file
+			"ffmpeg", "-i", inputFile,
+			// Convert to 1024kbps FLAC segment (ffmpeg defaults to 128kbps)
+			"-map", "0:a", "-c:a:0", "flac", "-b:a:0", "1024k",
```

Comment:
where does the 1024kbps come from?

Reply:
This was a bit arbitrary but I found this setting directly feeds into the bandwidth reported on the HLS stream metadata. I picked something grossly larger than the other format just so there's a clear disparity.

I'm not certain what the correct value of this should be. I did a test with a 96k 24b FLAC file as the source, converted it to this fmp4 HLS format using the default 128kbps, then downloaded the streamed HLS playback into a copy. The copy seems to have the same spectrograph as the original so I'm not quite sure how this setting works.

Reply:
Ok, I think this has something to do with compression: https://sound.stackexchange.com/questions/24680/why-does-a-flac-file-have-a-lower-bit-rate-than-the-cd-file

If I'm understanding that correctly, FLAC files can both preserve original quality + have a lower bitrate because of how FLAC does lossless compression. That would mean that it also depends on the type of sound contained in the FLAC file - if the content happens to be less-compressible, it's possible that the bitrate limit would impact quality.

This ffmpeg flag specifies the target bitrate for the encoding which can vary but I suppose it's best to treat it as an upper-bound with the case of FLAC. I might bump it up to 1,411kbps from https://www.gearpatrol.com/tech/audio/a36585957/lossless-audio-explained/ although this also depends on other factors (sample rate, etc.) but it's probably a decent enough ceiling.
```
The bitrate of a normal lossless audio track is around 1,411 kbps.
```